### PR TITLE
[ROCm] Allow bf16/bf16/fp32 in nvte_multi_tensor_gemm dispatcher

### PR DESCRIPTION
**Pull request by lizamd**
The is_supported_dtype check in nvte_multi_tensor_gemm previously required A==B==D for the fp16/bf16 path, which rejected the common bf16/bf16/fp32 case where the GEMM output is fp32 for gradient accumulation. This forced a fallback to multi_stream_cublas_gemm (a per-expert hipblaslt loop), bypassing the CK grouped GEMM kernel entirely on ROCm.

The CK FP16 dispatcher (ck_tile_grouped_gemm_fp16_dispatch) already supports independent D dtype via TRANSFORMER_ENGINE_TYPE_SWITCH_NON_FP8ONLY (fp32, fp16, bf16). The wrapper check is the only thing that prevents it from being reached.

Relaxed to require A==B in fp16/bf16 and D in {fp32, fp16, bf16}, which matches what the CK dispatcher actually accepts. Verified on Qwen3-30B-A3B MoE training on MI355X (gfx950): fallback warning rate drops from ~1040/step (every GEMM) to ~28/step (~3% of shapes that the CK kernel itself rejects via Kernel::IsSupportedArgument). Throughput is essentially unchanged in this workload because hipblaslt's per-shape autotuning happens to be competitive with the hardcoded CK tile configs for these MoE shapes; the gain will materialize once the CK dispatcher gains more tile configs (or shape-aware tile selection by aggregate M).

This is a CUDA path file; the same patch applies to the AMD path via hipify. No CUDA-side behavior change since cuBLAS/cutlass dispatch on NVIDIA still requires A==B==D in the cutlass fast-path pre-conditions.

# Description

Please include a brief summary of the changes, relevant motivation and context.

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Change A
- Change B

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by wenchenvincent** [comment]
@matthiasdiener @aris134 Could you review this PR?

**Review by wangye805 — CHANGES_REQUESTED** [review]
Can you edit an existing test or add a new test showing that with your change, bf16/fp16 input and fp32 outputs are going through the ck flow correctly now? Also paste some benchmarking data to this ticket for future reference

File: transformer_engine/common/gemm/cublaslt_gemm.cu [hunk 1]
```diff
@@ -1163,12 +1163,18 @@ void nvte_multi_tensor_gemm(const NVTETensor *A, const NVTETensor *B, NVTETensor
     auto A_dt = inputA->data.dtype;
     auto B_dt = inputB->data.dtype;
     auto D_dt = OutputD->data.dtype;
+    // CK FP16/BF16 grouped GEMM dispatcher (ck_tile_grouped_gemm_fp16_dispatch)
+    // already supports independent D dtype via TRANSFORMER_ENGINE_TYPE_SWITCH_NON_FP8ONLY
+    // (fp32, fp16, bf16). The previous check required A==B==D, which incorrectly
+    // rejected the common bf16/bf16/fp32 case (training with fp32 gradient
+    // accumulation), forcing a fallback to the per-expert hipblaslt loop.
+    // Relaxed to require A==B in fp16/bf16 and D in {fp32, fp16, bf16}.
```
Comment (aris134) [inline transformer_engine/common/gemm/cublaslt_gemm.cu]: I think this explanation may be better suited for the PR description rather than an inline code comment.

**Review by aris134 — CHANGES_REQUESTED** [review]
Agreed that the CK dispatch logic supports bf16/f32 combination. I would remove the detailed history comment about the previous fallback behavior which is better suited to the PR itself.

**Comment by lizamd** [comment]
AMD General


+ @Kang, ***@***.***>
Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by lizamd** [comment]
@wangye805 @aris134 could you check the new commit?

File: tests/pytorch/test_numerics.py [hunk 2]
```diff
@@ -3075,6 +3078,68 @@ def test_grouped_gemm(shape, dtype, layout, accumulate, use_cutlass):
         os.environ.pop("NVTE_USE_CUTLASS_GROUPED_GEMM", None)
 
 
+@pytest.mark.skipif(
+    torch.cuda.get_device_capability() != (9, 0) and not IS_HIP_EXTENSION,
+    reason="Only enable CUTLASS / CK grouped gemm on Hopper or ROCm",
+)
+@pytest.mark.parametrize("input_dtype", [torch.bfloat16, torch.float16])
+@pytest.mark.parametrize("layout", ["TN", "NT"])
+def test_grouped_gemm_fp32_output(input_dtype, layout):
```
Comment (ipanfilo) [inline tests/pytorch/test_numerics.py]: Can it be done by adding configs/parameters to test_grouped_gemm?

**Review by ipanfilo — CHANGES_REQUESTED** [review]

File: tests/pytorch/test_numerics.py [hunk 3]
```diff
@@ -3075,6 +3078,68 @@ def test_grouped_gemm(shape, dtype, layout, accumulate, use_cutlass):
...
+        grad = False
+    else:  # "NT" wgrad: weight gradient in fp32
+        A = list(torch.split(torch.randn(m, k, dtype=input_dtype, device="cuda"), m_splits))
+        B = list(torch.split(torch.randn(m, n, dtype=input_dtype, device="cuda"), m_splits))
+        out = [torch.empty(n, k, dtype=torch.float32, device="cuda") for _ in range(z)]
+        out_ref = [o.clone() for o in out]
+        single_output = False
+        grad = True
+
+    os.environ["NVTE_USE_CUTLASS_GROUPED_GEMM"] = "1"
```
Comment (wangye805) [inline tests/pytorch/test_numerics.py]: Besides Ilya's comment, please try to use monkeypatch for this env setting. You current setting is assuming user didn't set NVTE_USE_CUTLASS_GROUPED_GEMM when running pytests

**Review by wangye805 — CHANGES_REQUESTED** [review]

File: tests/pytorch/test_numerics.py [hunk 4]
```diff
@@ -2994,7 +2999,15 @@ def test_transformer_layer_hidden_states_format(dtype, bs, model):
 @pytest.mark.parametrize("layout", ["TN", "NN", "NT"])
 @pytest.mark.parametrize("accumulate", [False, True])
 @pytest.mark.parametrize("use_cutlass", use_cutlass_grouped_gemm)
-def test_grouped_gemm(shape, dtype, layout, accumulate, use_cutlass):
+@pytest.mark.parametrize("fp32_output", [False, True], ids=["out=input", "out=fp32"])
```
Comment (ipanfilo) [inline tests/pytorch/test_numerics.py:3002]: make if [False, True] is IS_HIP_EXTRENSION else [False,]
